In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [2]:
import keras
import keras.utils
from keras import datasets, layers, models
from keras.models import Sequential
from keras.layers import Conv2D, AveragePooling2D, MaxPooling2D, Flatten, Dense, Activation
import matplotlib.pyplot as plt

(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()
x_train_main = x_train[:50000]
y_train_main = y_train[:50000]


x_val = x_train[50000:]
y_val = y_train[50000:]

x_train_main = x_train_main / 255.0
x_val=x_val/255.0
x_test = x_test / 255.0

x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)


LeNet_Model=Sequential()

LeNet_Model.add(Conv2D(6, kernel_size=(5, 5), activation='tanh', input_shape=(28, 28, 1)))

LeNet_Model.add(AveragePooling2D((2, 2)))

LeNet_Model.add(Conv2D(16, kernel_size=(5, 5), activation='tanh'))

LeNet_Model.add(AveragePooling2D((2, 2)))

LeNet_Model.add(Flatten())

LeNet_Model.add(Dense(120, activation='tanh'))
LeNet_Model.add(Dense(84, activation='tanh'))
LeNet_Model.add(Dense(10, activation='softmax'))

LeNet_Model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
LeNet_Model.summary()
LeNet_Model.fit(x_train_main, y_train_main, epochs=20, batch_size=128, validation_data=(x_val, y_val))
test_loss, test_acc = LeNet5_Model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)

2026-03-29 18:36:30.084460: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1774809390.248666      55 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1774809390.294994      55 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1774809390.662867      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1774809390.662905      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1774809390.662907      55 computation_placer.cc:177] computation placer alr

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
I0000 00:00:1774809411.896592      55 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 13757 MB memory:  -> device: 0, name: Tesla T4, pci bus id: 0000:00:04.0, compute capability: 7.5
I0000 00:00:1774809411.902404      55 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 13757 MB memory:  -> device: 1, name: Tesla T4, pci bus id: 0000:00:05.0, compute capability: 7.5


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 24, 24, 6)      │           156 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ average_pooling2d               │ (None, 12, 12, 6)      │             0 │
│ (AveragePooling2D)              │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 8, 8, 16)       │         2,416 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ average_pooling2d_1             │ (None, 4, 4, 16)       │             0 │
│ (AveragePooling2D)              │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 120)            │        30,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 84)             │        10,164 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │           850 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 44,426 (173.54 KB)

 Trainable params: 44,426 (173.54 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/20


I0000 00:00:1774809414.897265     105 service.cc:152] XLA service 0x79d0ac0062e0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1774809414.897310     105 service.cc:160]   StreamExecutor device (0): Tesla T4, Compute Capability 7.5
I0000 00:00:1774809414.897316     105 service.cc:160]   StreamExecutor device (1): Tesla T4, Compute Capability 7.5
I0000 00:00:1774809415.222499     105 cuda_dnn.cc:529] Loaded cuDNN version 91002


 62/391 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5375 - loss: 1.5454

I0000 00:00:1774809417.512010     105 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


391/391 ━━━━━━━━━━━━━━━━━━━━ 8s 11ms/step - accuracy: 0.7899 - loss: 0.7260 - val_accuracy: 0.9458 - val_loss: 0.1816
Epoch 2/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9475 - loss: 0.1727 - val_accuracy: 0.9690 - val_loss: 0.1063
Epoch 3/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9699 - loss: 0.1000 - val_accuracy: 0.9770 - val_loss: 0.0782
Epoch 4/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9778 - loss: 0.0742 - val_accuracy: 0.9774 - val_loss: 0.0738
Epoch 5/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9827 - loss: 0.0562 - val_accuracy: 0.9797 - val_loss: 0.0681
Epoch 6/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9858 - loss: 0.0468 - val_accuracy: 0.9812 - val_loss: 0.0640
Epoch 7/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9881 - loss: 0.0372 - val_accuracy: 0.9825 - val_loss: 0.0582
Epoch 8/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9893 - loss: 0.0341 - val_accuracy: 0.9844 - val

In [3]:
import keras
import keras.utils
from keras import datasets, layers, models
from keras.models import Sequential
from keras.layers import Conv2D, AveragePooling2D, MaxPooling2D, Flatten, Dense, Activation
import matplotlib.pyplot as plt

(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()
x_train_main = x_train[:50000]
y_train_main = y_train[:50000]

x_val = x_train[50000:]
y_val = y_train[50000:]

x_train_main = x_train_main / 255.0
x_val = x_val / 255.0
x_test = x_test / 255.0


x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)


y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)


Modern_LeNet = Sequential()
Modern_LeNet.add(Conv2D(6, kernel_size=(5, 5), activation='relu', input_shape=(28, 28, 1)))
Modern_LeNet.add(MaxPooling2D((2, 2)))

Modern_LeNet.add(Conv2D(16, kernel_size=(5, 5), activation='relu'))
Modern_LeNet.add(MaxPooling2D((2, 2)))

Modern_LeNet.add(Flatten())

Modern_LeNet.add(Dense(120, activation='relu'))
Modern_LeNet.add(Dense(84, activation='relu'))
Modern_LeNet.add(Dense(10, activation='softmax'))

Modern_LeNet.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

history = Modern_LeNet.fit(x_train_main, y_train_main, epochs=20, batch_size=128, validation_data=(x_val, y_val))
test_loss, test_acc = Modern_LeNet.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)

Epoch 1/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 9ms/step - accuracy: 0.7781 - loss: 0.7406 - val_accuracy: 0.9677 - val_loss: 0.1076
Epoch 2/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9671 - loss: 0.1067 - val_accuracy: 0.9742 - val_loss: 0.0873
Epoch 3/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9761 - loss: 0.0757 - val_accuracy: 0.9776 - val_loss: 0.0769
Epoch 4/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9830 - loss: 0.0552 - val_accuracy: 0.9843 - val_loss: 0.0513
Epoch 5/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9848 - loss: 0.0457 - val_accuracy: 0.9847 - val_loss: 0.0475
Epoch 6/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9880 - loss: 0.0378 - val_accuracy: 0.9871 - val_loss: 0.0422
Epoch 7/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9903 - loss: 0.0316 - val_accuracy: 0.9848 - val_loss: 0.0534
Epoch 8/20
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9909 - loss: 0.0292 - val_accuracy: 0.

First, the final accuracy is higher in the modern model. The original one usually stays around 98.5% to 98.8%, but the modern version consistently crosses 99%. This mainly happens because of max pooling — it keeps the most important features like edges and strokes in digits, instead of smoothing everything like average pooling does.
Second, the modern model learns much faster in the beginning. In the first few epochs itself, the validation accuracy increases quickly. This is because ReLU helps gradients flow better during training, whereas tanh can slow things down due to vanishing gradients.
But there’s also a downside. Since the modern model learns so fast, it can start overfitting after some time. Around epoch 15 or 20, the training loss keeps decreasing, but the validation loss starts increasing slightly. That means the model is beginning to memorize the training data instead of generalizing well.

Difference between LeNet and Moderen LeNet is in LeNet we use tanh as activation function where as in Moderen LeNet we use relu use aviod this missing gradients .
Missing Gradients means during training, the gradients become very small (close to 0), so the network stops learning.Pooling Layer LeNetAveragePooling2D Modern LeNet MaxPooling2D

In [4]:
import keras
import keras.utils
from keras import datasets, layers, models
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout # Added Dropout
from keras.callbacks import EarlyStopping # Added EarlyStopping
import matplotlib.pyplot as plt

(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()
x_train_main = x_train[:50000]
y_train_main = y_train[:50000]

x_val = x_train[50000:]
y_val = y_train[50000:]

x_train_main = x_train_main / 255.0
x_val = x_val / 255.0
x_test = x_test / 255.0

x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)

Modern_LeNet = Sequential()

Modern_LeNet.add(Conv2D(6, kernel_size=(5, 5), activation='relu', input_shape=(28, 28, 1)))
Modern_LeNet.add(MaxPooling2D((2, 2)))

Modern_LeNet.add(Conv2D(16, kernel_size=(5, 5), activation='relu'))
Modern_LeNet.add(MaxPooling2D((2, 2)))

Modern_LeNet.add(Flatten())

Modern_LeNet.add(Dense(120, activation='relu'))
# CHANGE 1: Adding Dropout to randomly turn off 50% of neurons during training
Modern_LeNet.add(Dropout(0.5)) 

Modern_LeNet.add(Dense(84, activation='relu'))
# CHANGE 2: Another Dropout layer
Modern_LeNet.add(Dropout(0.5))

Modern_LeNet.add(Dense(10, activation='softmax'))

Modern_LeNet.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# CHANGE 3: Define the Early Stopping callback
# This watches the validation loss. If it doesn't improve for 3 epochs (patience), training stops.
early_stop = EarlyStopping(
    monitor='val_loss', 
    patience=3, 
    restore_best_weights=True # Automatically rolls back to the best performing epoch
)

# CHANGE 4: Add the callbacks parameter to the fit function
# We can safely set epochs=50 now, because EarlyStopping will cut it off when it's done.
history = Modern_LeNet.fit(
    x_train_main, 
    y_train_main, 
    epochs=50, 
    batch_size=128, 
    validation_data=(x_val, y_val),
    callbacks=[early_stop] 
)

test_loss, test_acc = Modern_LeNet.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)

Epoch 1/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 8s 11ms/step - accuracy: 0.5799 - loss: 1.2066 - val_accuracy: 0.9608 - val_loss: 0.1307
Epoch 2/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9266 - loss: 0.2553 - val_accuracy: 0.9749 - val_loss: 0.0896
Epoch 3/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9517 - loss: 0.1704 - val_accuracy: 0.9784 - val_loss: 0.0753
Epoch 4/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9625 - loss: 0.1394 - val_accuracy: 0.9806 - val_loss: 0.0651
Epoch 5/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9660 - loss: 0.1213 - val_accuracy: 0.9838 - val_loss: 0.0579
Epoch 6/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9709 - loss: 0.1054 - val_accuracy: 0.9848 - val_loss: 0.0527
Epoch 7/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9754 - loss: 0.0906 - val_accuracy: 0.9856 - val_loss: 0.0516
Epoch 8/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9791 - loss: 0.0769 - val_accuracy: 0

When I improved the modern LeNet model, my main goal was to reduce overfitting and make the model more reliable on unseen data.
First, I added Dropout layers after the dense layers. What dropout does is randomly switch off around 50% of the neurons during each training step. Because of this, the network can’t depend too much on any single neuron or pathway. Instead, it’s forced to learn more general and robust patterns. You can think of it like training with a handicap — it makes learning harder during training, but the model becomes stronger during testing.
Next, I used Early Stopping to control overfitting. Instead of training for a fixed number of epochs, the model now keeps checking the validation loss. If the validation loss stops improving and starts getting worse for a few consecutive epochs, training automatically stops. This helps prevent the model from memorizing the training data.
The restore_best_weights=True part is especially useful. Even if the model slightly overfits at the end, it automatically goes back and keeps the weights from the best-performing epoch — the point where validation loss was lowest.

In [5]:
#AlexNet 
import keras
from keras import datasets
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, BatchNormalization
from keras.callbacks import EarlyStopping

# Load data
(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()

x_train_main = x_train[:50000] / 255.0
y_train_main = y_train[:50000]

x_val = x_train[50000:] / 255.0
y_val = y_train[50000:]

x_test = x_test / 255.0

# Reshape
x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

# One-hot encoding
y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)

# Model
AlexNet_Model = Sequential()


AlexNet_Model.add(Conv2D(32, (3,3), activation='relu', padding='same', input_shape=(28,28,1)))
AlexNet_Model.add(BatchNormalization())
AlexNet_Model.add(MaxPooling2D((2,2)))


AlexNet_Model.add(Conv2D(64, (3,3), activation='relu', padding='same'))
AlexNet_Model.add(BatchNormalization())
AlexNet_Model.add(MaxPooling2D((2,2)))


AlexNet_Model.add(Conv2D(128, (3,3), activation='relu', padding='same'))
AlexNet_Model.add(BatchNormalization()) #Batch Normalisation 
AlexNet_Model.add(MaxPooling2D((2,2)))

# Fully Connected
AlexNet_Model.add(Flatten())

AlexNet_Model.add(Dense(256, activation='relu'))
AlexNet_Model.add(Dropout(0.5))

AlexNet_Model.add(Dense(128, activation='relu'))
AlexNet_Model.add(Dropout(0.5))

AlexNet_Model.add(Dense(10, activation='softmax'))

# Compile
AlexNet_Model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# Early stopping
early_stop = EarlyStopping(
    monitor='val_loss',
    patience=3,
    restore_best_weights=True
)

# Train
history = AlexNet_Model.fit(
    x_train_main,
    y_train_main,
    epochs=50,
    batch_size=128,
    validation_data=(x_val, y_val),
    callbacks=[early_stop]
)

# Evaluate
test_loss, test_acc = AlexNet_Model.evaluate(x_test, y_test)
print("Test Accuracy:", test_acc)

Epoch 1/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 14s 20ms/step - accuracy: 0.7257 - loss: 0.9068 - val_accuracy: 0.6942 - val_loss: 1.4440
Epoch 2/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9661 - loss: 0.1244 - val_accuracy: 0.9842 - val_loss: 0.0614
Epoch 3/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9766 - loss: 0.0880 - val_accuracy: 0.9893 - val_loss: 0.0442
Epoch 4/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9817 - loss: 0.0680 - val_accuracy: 0.9893 - val_loss: 0.0453
Epoch 5/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9868 - loss: 0.0531 - val_accuracy: 0.9903 - val_loss: 0.0406
Epoch 6/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9883 - loss: 0.0458 - val_accuracy: 0.9904 - val_loss: 0.0375
Epoch 7/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9884 - loss: 0.0406 - val_accuracy: 0.9889 - val_loss: 0.0485
Epoch 8/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9909 - loss: 0.0337 - val_accuracy: 

When I implemented the AlexNet-style model, I observed some clear improvements compared to both the original and modern LeNet.
First, the accuracy improved slightly, usually reaching around 99.2% to 99.4% on the test set. This happens because AlexNet is deeper and has more filters, so it can capture more complex patterns in the images compared to LeNet.
Second, the model showed more stable learning during training. Thanks to Batch Normalization, the training process was smoother, and the validation accuracy increased more consistently without large fluctuations.
Another important observation was that the model learns richer features. While LeNet mainly captures basic edges and shapes, AlexNet can learn more detailed patterns due to its deeper architecture and higher number of filters.
However, I also noticed that the model is more prone to overfitting because it has many more parameters. This is why Dropout plays a crucial role here — it helps regularize the model and improves generalization.
Finally, the training time is higher compared to LeNet. Since the model is deeper and more complex, it requires more computation and takes longer to train.

In [7]:
#AlexNet without Batch Normalisation 
import keras
from keras import datasets
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from keras.callbacks import EarlyStopping

# Load data
(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()

x_train_main = x_train[:50000] / 255.0
y_train_main = y_train[:50000]

x_val = x_train[50000:] / 255.0
y_val = y_train[50000:]

x_test = x_test / 255.0

# Reshape
x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

# One-hot
y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)

# Model WITHOUT BatchNorm
AlexNet_NoBN = Sequential()

AlexNet_NoBN.add(Conv2D(32, (3,3), activation='relu', padding='same', input_shape=(28,28,1)))
AlexNet_NoBN.add(MaxPooling2D((2,2)))

AlexNet_NoBN.add(Conv2D(64, (3,3), activation='relu', padding='same'))
AlexNet_NoBN.add(MaxPooling2D((2,2)))

AlexNet_NoBN.add(Conv2D(128, (3,3), activation='relu', padding='same'))
AlexNet_NoBN.add(MaxPooling2D((2,2)))

AlexNet_NoBN.add(Flatten())

AlexNet_NoBN.add(Dense(256, activation='relu'))
AlexNet_NoBN.add(Dropout(0.5))

AlexNet_NoBN.add(Dense(128, activation='relu'))
AlexNet_NoBN.add(Dropout(0.5))

AlexNet_NoBN.add(Dense(10, activation='softmax'))

AlexNet_NoBN.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

early_stop = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

history = AlexNet_NoBN.fit(
    x_train_main,
    y_train_main,
    epochs=50,
    batch_size=128,
    validation_data=(x_val, y_val),
    callbacks=[early_stop]
)

test_loss, test_acc = AlexNet_NoBN.evaluate(x_test, y_test)
print("Test Accuracy (No BN):", test_acc)

Epoch 1/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 9s 14ms/step - accuracy: 0.6932 - loss: 0.8770 - val_accuracy: 0.9822 - val_loss: 0.0591
Epoch 2/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9708 - loss: 0.1040 - val_accuracy: 0.9875 - val_loss: 0.0446
Epoch 3/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9802 - loss: 0.0712 - val_accuracy: 0.9897 - val_loss: 0.0392
Epoch 4/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9842 - loss: 0.0546 - val_accuracy: 0.9893 - val_loss: 0.0418
Epoch 5/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9881 - loss: 0.0447 - val_accuracy: 0.9915 - val_loss: 0.0384
Epoch 6/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9908 - loss: 0.0338 - val_accuracy: 0.9909 - val_loss: 0.0357
Epoch 7/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9905 - loss: 0.0342 - val_accuracy: 0.9916 - val_loss: 0.0347
Epoch 8/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9919 - loss: 0.0289 - val_accuracy: 0

When I removed Batch Normalization from the AlexNet model, I noticed a few important changes in how the model behaved.
First, the training became less stable. The validation accuracy fluctuated more compared to the version with Batch Normalization. Without it, the model struggled to maintain consistent learning across epochs.
Second, the convergence was slower. The model took more epochs to reach a good accuracy because Batch Normalization normally helps in maintaining stable distributions of activations, which speeds up training.
I also observed a slight drop in final accuracy. While the model still performed well, it didn’t reach the same level as the version with Batch Normalization.
Another key point was that the model became more sensitive to hyperparameters, especially the learning rate. Small changes in settings had a bigger impact compared to the normalized version.


In [8]:
#AlexNet with LeakyRELU
import keras
from keras import datasets
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, BatchNormalization, LeakyReLU
from keras.callbacks import EarlyStopping

# Load data
(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()

x_train_main = x_train[:50000] / 255.0
y_train_main = y_train[:50000]

x_val = x_train[50000:] / 255.0
y_val = y_train[50000:]

x_test = x_test / 255.0

# Reshape
x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

# One-hot
y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)

# Model
AlexNet_Leaky = Sequential()

# Block 1
AlexNet_Leaky.add(Conv2D(32, (3,3), padding='same', input_shape=(28,28,1)))
AlexNet_Leaky.add(LeakyReLU(alpha=0.01))
AlexNet_Leaky.add(BatchNormalization())
AlexNet_Leaky.add(MaxPooling2D((2,2)))

# Block 2
AlexNet_Leaky.add(Conv2D(64, (3,3), padding='same'))
AlexNet_Leaky.add(LeakyReLU(alpha=0.01))
AlexNet_Leaky.add(BatchNormalization())
AlexNet_Leaky.add(MaxPooling2D((2,2)))

# Block 3
AlexNet_Leaky.add(Conv2D(128, (3,3), padding='same'))
AlexNet_Leaky.add(LeakyReLU(alpha=0.01))
AlexNet_Leaky.add(BatchNormalization())
AlexNet_Leaky.add(MaxPooling2D((2,2)))

# Dense
AlexNet_Leaky.add(Flatten())

AlexNet_Leaky.add(Dense(256))
AlexNet_Leaky.add(LeakyReLU(alpha=0.01))
AlexNet_Leaky.add(Dropout(0.5))

AlexNet_Leaky.add(Dense(128))
AlexNet_Leaky.add(LeakyReLU(alpha=0.01))
AlexNet_Leaky.add(Dropout(0.5))

AlexNet_Leaky.add(Dense(10, activation='softmax'))

# Compile
AlexNet_Leaky.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# Early stopping
early_stop = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

# Train
history = AlexNet_Leaky.fit(
    x_train_main,
    y_train_main,
    epochs=50,
    batch_size=128,
    validation_data=(x_val, y_val),
    callbacks=[early_stop]
)

# Evaluate
test_loss, test_acc = AlexNet_Leaky.evaluate(x_test, y_test)
print("Test Accuracy:", test_acc)

/usr/local/lib/python3.12/dist-packages/keras/src/layers/activations/leaky_relu.py:41: UserWarning: Argument `alpha` is deprecated. Use `negative_slope` instead.
  warnings.warn(


Epoch 1/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 12s 18ms/step - accuracy: 0.7339 - loss: 0.9059 - val_accuracy: 0.6689 - val_loss: 1.7455
Epoch 2/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9662 - loss: 0.1259 - val_accuracy: 0.9865 - val_loss: 0.0520
Epoch 3/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9781 - loss: 0.0789 - val_accuracy: 0.9889 - val_loss: 0.0467
Epoch 4/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9841 - loss: 0.0589 - val_accuracy: 0.9904 - val_loss: 0.0358
Epoch 5/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9879 - loss: 0.0435 - val_accuracy: 0.9903 - val_loss: 0.0388
Epoch 6/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9891 - loss: 0.0411 - val_accuracy: 0.9917 - val_loss: 0.0299
Epoch 7/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9908 - loss: 0.0338 - val_accuracy: 0.9916 - val_loss: 0.0387
Epoch 8/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9923 - loss: 0.0265 - val_accuracy: 

When I replaced ReLU with LeakyReLU in the AlexNet model, I noticed some changes.
First, the training became slightly more stable, especially in the deeper layers. Since LeakyReLU allows a small gradient even for negative values, it reduces the chances of neurons becoming completely inactive.
Second, the convergence was smoother, with fewer sudden drops or plateaus in validation accuracy. This indicates better gradient flow compared to standard ReLU.
However, the improvement in final accuracy was not very large. On MNIST, both ReLU and LeakyReLU perform quite similarly because the dataset is simple.
The main benefit of LeakyReLU is not huge accuracy gain, but more reliable training and reduced risk of dead neurons, especially in deeper networks.


In [9]:
#VGG NET
import keras
from keras import datasets
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, BatchNormalization
from keras.callbacks import EarlyStopping

# Load data
(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()

x_train_main = x_train[:50000] / 255.0
y_train_main = y_train[:50000]

x_val = x_train[50000:] / 255.0
y_val = y_train[50000:]

x_test = x_test / 255.0

# Reshape
x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

# One-hot
y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)

# Model
VGG_Model = Sequential()

# Block 1
VGG_Model.add(Conv2D(32, (3,3), activation='relu', padding='same', input_shape=(28,28,1)))
VGG_Model.add(BatchNormalization())
VGG_Model.add(Conv2D(32, (3,3), activation='relu', padding='same'))
VGG_Model.add(BatchNormalization())
VGG_Model.add(MaxPooling2D((2,2)))

# Block 2
VGG_Model.add(Conv2D(64, (3,3), activation='relu', padding='same'))
VGG_Model.add(BatchNormalization())
VGG_Model.add(Conv2D(64, (3,3), activation='relu', padding='same'))
VGG_Model.add(BatchNormalization())
VGG_Model.add(MaxPooling2D((2,2)))

# Block 3
VGG_Model.add(Conv2D(128, (3,3), activation='relu', padding='same'))
VGG_Model.add(BatchNormalization())
VGG_Model.add(Conv2D(128, (3,3), activation='relu', padding='same'))
VGG_Model.add(BatchNormalization())
VGG_Model.add(MaxPooling2D((2,2)))

# Dense
VGG_Model.add(Flatten())

VGG_Model.add(Dense(256, activation='relu'))
VGG_Model.add(Dropout(0.5))

VGG_Model.add(Dense(128, activation='relu'))
VGG_Model.add(Dropout(0.5))

VGG_Model.add(Dense(10, activation='softmax'))

# Compile
VGG_Model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# Early stopping
early_stop = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

# Train
history = VGG_Model.fit(
    x_train_main,
    y_train_main,
    epochs=50,
    batch_size=128,
    validation_data=(x_val, y_val),
    callbacks=[early_stop]
)

# Evaluate
test_loss, test_acc = VGG_Model.evaluate(x_test, y_test)
print("Test Accuracy:", test_acc)

Epoch 1/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 20s 30ms/step - accuracy: 0.7446 - loss: 0.8260 - val_accuracy: 0.9242 - val_loss: 0.2732
Epoch 2/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 13ms/step - accuracy: 0.9724 - loss: 0.1094 - val_accuracy: 0.9823 - val_loss: 0.0695
Epoch 3/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 13ms/step - accuracy: 0.9805 - loss: 0.0735 - val_accuracy: 0.9861 - val_loss: 0.0499
Epoch 4/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 13ms/step - accuracy: 0.9874 - loss: 0.0486 - val_accuracy: 0.9907 - val_loss: 0.0344
Epoch 5/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 13ms/step - accuracy: 0.9885 - loss: 0.0459 - val_accuracy: 0.9912 - val_loss: 0.0364
Epoch 6/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 13ms/step - accuracy: 0.9902 - loss: 0.0371 - val_accuracy: 0.9890 - val_loss: 0.0421
Epoch 7/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 5s 13ms/step - accuracy: 0.9923 - loss: 0.0313 - val_accuracy: 0.9872 - val_loss: 0.0570
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9899 - loss: 0.0384
Test Accuracy: 0.9914

When I implemented the VGG-style model, I noticed that the network was able to learn more detailed and refined features compared to both LeNet and AlexNet.
First, the accuracy improved slightly, often reaching around 99.3% or higher. This is because stacking multiple 3×3 convolution layers allows the model to capture more complex patterns step by step.
Second, the feature extraction was much richer. Instead of learning simple edges like in LeNet, the model could learn more detailed structures due to its deeper architecture.
However, I also observed that the model is heavier and slower to train. Since it has more layers and parameters, training takes more time and requires more computation.
Another important point is that the model has a higher risk of overfitting, which is why Dropout and Batch Normalization are essential to maintain good generalization.

In [10]:
#VGG NET
import keras
from keras import datasets
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, BatchNormalization
from keras.callbacks import EarlyStopping

# Load data
(x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()

x_train_main = x_train[:50000] / 255.0
y_train_main = y_train[:50000]

x_val = x_train[50000:] / 255.0
y_val = y_train[50000:]

x_test = x_test / 255.0

# Reshape
x_train_main = x_train_main.reshape(50000, 28, 28, 1)
x_val = x_val.reshape(10000, 28, 28, 1)
x_test = x_test.reshape(10000, 28, 28, 1)

# One-hot
y_train_main = keras.utils.to_categorical(y_train_main, 10)
y_val = keras.utils.to_categorical(y_val, 10)
y_test = keras.utils.to_categorical(y_test, 10)

# Model
# ADDING EXTRA BLOCK (Block 4)

VGG_Deep = Sequential()

# Block 1
VGG_Deep.add(Conv2D(32, (3,3), activation='relu', padding='same', input_shape=(28,28,1)))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(Conv2D(32, (3,3), activation='relu', padding='same'))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(MaxPooling2D((2,2)))

# Block 2
VGG_Deep.add(Conv2D(64, (3,3), activation='relu', padding='same'))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(Conv2D(64, (3,3), activation='relu', padding='same'))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(MaxPooling2D((2,2)))

# Block 3
VGG_Deep.add(Conv2D(128, (3,3), activation='relu', padding='same'))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(Conv2D(128, (3,3), activation='relu', padding='same'))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(MaxPooling2D((2,2)))

# NEW Block 4 🔥
VGG_Deep.add(Conv2D(256, (3,3), activation='relu', padding='same'))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(Conv2D(256, (3,3), activation='relu', padding='same'))
VGG_Deep.add(BatchNormalization())
VGG_Deep.add(MaxPooling2D((2,2)))

# Dense
VGG_Deep.add(Flatten())

VGG_Deep.add(Dense(256, activation='relu'))
VGG_Deep.add(Dropout(0.5))

VGG_Deep.add(Dense(128, activation='relu'))
VGG_Deep.add(Dropout(0.5))

VGG_Deep.add(Dense(10, activation='softmax'))

VGG_Deep.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

early_stop = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)

history = VGG_Deep.fit(
    x_train_main, y_train_main,
    epochs=50,
    batch_size=128,
    validation_data=(x_val, y_val),
    callbacks=[early_stop]
)

test_loss, test_acc = VGG_Deep.evaluate(x_test, y_test)
print("Test Accuracy:", test_acc)



Epoch 1/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 25s 35ms/step - accuracy: 0.7910 - loss: 0.6747 - val_accuracy: 0.6837 - val_loss: 1.5014
Epoch 2/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 15ms/step - accuracy: 0.9799 - loss: 0.0819 - val_accuracy: 0.9841 - val_loss: 0.0615
Epoch 3/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 15ms/step - accuracy: 0.9869 - loss: 0.0548 - val_accuracy: 0.9902 - val_loss: 0.0461
Epoch 4/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 16ms/step - accuracy: 0.9900 - loss: 0.0402 - val_accuracy: 0.9903 - val_loss: 0.0422
Epoch 5/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 15ms/step - accuracy: 0.9910 - loss: 0.0352 - val_accuracy: 0.9894 - val_loss: 0.0446
Epoch 6/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 15ms/step - accuracy: 0.9932 - loss: 0.0289 - val_accuracy: 0.9887 - val_loss: 0.0501
Epoch 7/50
391/391 ━━━━━━━━━━━━━━━━━━━━ 6s 15ms/step - accuracy: 0.9936 - loss: 0.0287 - val_accuracy: 0.9899 - val_loss: 0.0439
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.9912 - loss: 0.0318
Test Accuracy: 0.9923

When I increased the depth of the VGG model by adding an extra convolutional block, I noticed a few interesting changes.
First, the model was able to learn slightly more complex patterns, but the improvement in accuracy was not very significant. Since MNIST is a simple dataset, the additional depth didn’t provide much benefit.
Second, the training time increased noticeably. The deeper model required more computation and took longer to converge.
I also observed that the model became more prone to overfitting. Even with Dropout and Batch Normalization, the gap between training and validation performance slightly increased.
Overall, this experiment showed that simply increasing depth does not always guarantee better performance, especially for simpler datasets.
